application/common/components/Emailer.php

```diff
@@ -694,24 +703,36 @@ public function sendPasswordExpiredEmails()
             return;
         }
 
+        $logData = [
+            'action' => 'send password expired notices',
+            'status' => 'starting',
+        ];
+
+        $users = User::getActiveUnlockedUsers();
+
+        $this->logger->info(array_merge($logData, [
+            'active_users' => count($users)
+        ]));
+
         $numEmailsSent = 0;
-        $users = User::findAll(['active' => 'yes', 'locked' => 'no', ]);
         foreach ($users as $user) {
             /** @var Password $userPassword */
             $userPassword = $user->currentPassword;
-            if ($userPassword
-                && strtotime($userPassword->getExpiresOn()) < time()
-                && ! $this->hasReceivedMessageRecently($user->id, EmailLog::MESSAGE_TYPE_PASSWORD_EXPIRED)
-            ) {
-                $this->sendMessageTo(EmailLog::MESSAGE_TYPE_PASSWORD_EXPIRED, $user);
-                $numEmailsSent++;
+            if ($userPassword) {
+                $passwordExpiry = strtotime($userPassword->getExpiresOn());
+                if ($passwordExpiry < time()
+                    && $passwordExpiry > strtotime('-15 days')
```

Comment:
Not a big deal and maybe not worth it at this stage.  I was thinking these strtotime values (here and line 681) should probably be constants and maybe even configurable with a default.  (If we make them configurable, though, we should probably include a validator that ensures there's no typo or something that could sabotage it.)

Reply:
I'll change to constants, but don't want to mess with config vars at this point.